internal/testing/exchange/exchange.go

```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		return fmt.Errorf("mock server error %s", err)
+	}
+	b := e.GetBase()
+	b.SkipAuthCheck = true
+	err = b.SetHTTPClient(newClient)
+	if err != nil {
+		return fmt.Errorf("mock server error %s", err)
+	}
+	endpointMap := b.API.Endpoints.GetURLMap()
+	for k := range endpointMap {
+		err = b.API.Endpoints.SetRunning(k, serverDetails)
+		if err != nil {
+			return fmt.Errorf("mock server error %s", err)
+		}
+	}
+	log.Printf(sharedtestvalues.MockTesting, e.GetName())
+
+	return nil
+}
+
+var upgrader = websocket.Upgrader{}
+
+type wsMockFunc func(msg []byte, w *websocket.Conn) error
+
+// MockWSInstance creates a new Exchange instance with a mock WS instance and HTTP server
+// It accepts an exchange package type argument and a mock WS function
+// It is expected to be run from any WS tests which need a specific response function
+func MockWSInstance[T any, PT interface {
+	*T
+	exchange.IBotExchange
+}](tb testing.TB, m wsMockFunc) *T {
+	tb.Helper()
+
+	e := PT(new(T))
+	require.NoError(tb, TestInstance(e), "TestInstance setup should not error")
+
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { wsMockWrapper(tb, w, r, m) }))
+
+	b := e.GetBase()
+	b.SkipAuthCheck = true
+	err := b.API.Endpoints.SetRunning("RestSpotURL", s.URL)
+	require.NoError(tb, err, "Endpoints.SetRunning should not error for RestSpotURL")
+	for _, auth := range []bool{true, false} {
+		err = b.Websocket.SetWebsocketURL("ws"+strings.TrimPrefix(s.URL, "http"), auth, true)
+		require.NoErrorf(tb, err, "SetWebsocketURL should not error for auth: %v", auth)
+	}
+
+	b.Features.Subscriptions = []*subscription.Subscription{}
+	err = b.Websocket.Connect()
+	require.NoError(tb, err, "Connect should not error")
+
+	return e
+}
+
+// wsMockWrapper handles upgrading an initial HTTP request to WS, and then runs a for loop calling the mock func on each input
+func wsMockWrapper(tb testing.TB, w http.ResponseWriter, r *http.Request, m wsMockFunc) {
+	tb.Helper()
+	// TODO: This needs to move once this branch includes #1358, probably to use a new mock HTTP instance for kraken
+	if strings.Contains(r.URL.Path, "GetWebSocketsToken") {
+		_, err := w.Write([]byte(`{"result":{"token":"mockAuth"}}`))
+		require.NoError(tb, err, "Write should not error")
+		return
+	}
+	c, err := upgrader.Upgrade(w, r, nil)
+	require.NoError(tb, err, "Upgrade connection should not error")
+	defer c.Close()
+	for {
+		_, p, err := c.ReadMessage()
+		require.NoError(tb, err, "ReadMessage should not error")
+
+		err = m(p, c)
+		assert.NoError(tb, err, "WS Mock Function should not error")
+	}
+}
+
+var setupWsMutex sync.Mutex
+var setupWsOnce = make(map[exchange.IBotExchange]bool)
+
+// SetupWs is a helper function to connect both auth and normal websockets
+// It will skip the test if websockets are not enabled
+// It's up to the test to skip if it requires creds, though
+func SetupWs(tb testing.TB, e exchange.IBotExchange) {
```

Comment:
Yep. [See this comment](https://github.com/thrasher-corp/gocryptotrader/pull/1394#issuecomment-1894908383)

I brought the shared testing framework stuff in from another branch to fix one of the issues with kucoin.

Is that okay?

I REALLY don't want to isolate parts of this branch I pulled in. I'll be using these shared methods across all exchanges in tests, with kraken and binance already done in those branches.

Reply:
Nah Its fine, I was just highlighting it. Unused code doesn't matter so much here since I know with all your various rebases between PRs and such that it will be used